### DISCUSSION TITLE
HL-ZASM Preprocessor

### DISCUSSION THREAD

**Discussion by suunr** [wiremod/wire-cpu#7] [General]
It would be better if the preprocessor should be executed before any compilation process and **not** during it. That means that the preprocessor tries to resolve the label, which leads to errors.

``` C
#define MY_FUNCTION some_func
[...]
```

An ANSI-C preprocessor would replace all `MY_FUNCTION`s in the source code with `some_func`. It doesn't throw an error if a function or variable called `some_func` doesn't exist.
However the HL-ZASM compiler throws an error. This can be annoying if you want to have all `#define`s before the actual source code.

This idea is taken from an ANSI-C compiler, because the HL-ZASM preprocessor is just poor.

**Comment by AbigailBuccaneer** [comment]
Is there any case in which the HL-ZASM preprocessor accepts things that the C preprocessor doesn't, or cases in which they both accept it but behave differently?

**Comment by suunr** [comment]
The HL-ZASM preprocessor has some specific pragma commands (pragma is compiler-specific anyway). Except that there should be nothing which the C preprocessor doesn't have.

**Comment by suunr** [comment]
What about this https://github.com/willsteel/lcpp? Maybe we could embed it in the compiler.

**Comment by oldmud0** [comment]
@suunrider Are you sure this library doesn't depend on the presence of LuaJIT?
Hmm... it's just short of 2000 lines.. MIT license, that's compatible.. looks pretty stable... a bit concerned about how `__FILE__` would be handled though...

But other than that, I think I'll :white_check_mark: support this idea

**Comment by suunr** [comment]
It works, but there are 2 problems: It doesn't properly check if the first character of line is "#", instead it looks for every single character. The other problem is that it takes much time when compiling >200 lines of code (maybe because it checks for every single character). Setting `__FILE__` isn't a big deal

**Comment by suunr** [comment]
I fixed the first problem, also found out it fixes #619.
